lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,303 @@
+		'onsite' => array(
+			'code' => 'onsite',
+			'internalcode'=> 'onsite',
+			'label'=> 'Collect data locally',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '0',
+			'required'=> true,
+		),
+		'cancelUrl' => array(
+			'code' => 'cancelurl',
+			'internalcode'=> 'cancelurl',
+			'label'=> 'cancelurl for paypal',
+			'type'=> 'string',
+			'internaltype'=> 'string',
+			'default'=> 'www.example.com',
+			'required'=> false,
+		),
+		'address' => array(
+			'code' => 'address',
+			'internalcode'=> 'address',
+			'label'=> 'Send address to payment gateway too',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '0',
+			'required'=> false,
+		),
+		'authorize' => array(
+			'code' => 'authorize',
+			'internalcode'=> 'authorize',
+			'label'=> 'Authorize payments and capture later',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '1',
+			'required'=> true,
+		),
+		'createtoken' => array(
+			'code' => 'createtoken',
+			'internalcode'=> 'createtoken',
+			'label'=> 'Request token for recurring payments',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '1',
+			'required'=> true,
+		),
+		'testmode' => array(
+			'code' => 'testmode',
+			'internalcode'=> 'testmode',
+			'label'=> 'Test mode without payments',
+			'type'=> 'boolean',
+			'internaltype'=> 'boolean',
+			'default'=> '0',
+			'required'=> true,
+		),
+	);
+	private $provider;
+
+	private $feConfig = array(
+		'payment.firstname' => array(
+			'code' => 'payment.firstname',
+			'internalcode'=> 'firstName',
+			'label'=> 'First name',
+			'type'=> 'string',
+			'internaltype'=> 'string',
+			'default'=> '',
+		//	'public' => false ,
+ 			'required'=> false
+		),
+	);
+
+	public function getConfigBE() : array
+	{
+		$list = [];
+
+		foreach( $this->beConfig as $key => $config ) {
+			$list[$key] = new \Aimeos\MW\Criteria\Attribute\Standard( $config );
+		}
+
+		return $list;
+	}
+
+	protected function getProvider() : \Omnipay\Common\GatewayInterface
+	{
+		if( !isset( $this->provider ) )
+		{
+			$this->provider = OPay::create( 'PayPal_Rest' );
+			$this->provider->setTestMode( (bool) $this->getValue( 'testmode', false ) );
+			$this->provider->initialize( $this->getServiceItem()->getConfig() );
+		}
+
+		return $this->provider;
+	}
+
+
+	protected function sendRequest( \Aimeos\MShop\Order\Item\Iface $order, array $data ) : \Omnipay\Common\Message\ResponseInterface
+	{
+		$provider = $this->getProvider();
+
+		if( $this->getValue( 'authorize', false ) && $provider->supportsAuthorize() )
+		{
+			$response = $provider->authorize( $data )->send();
+			$order->setPaymentStatus( Status::PAY_AUTHORIZED );
+		}
+		else
+		{
+			$response = $provider->purchase( $data )->send();
+			$order->setPaymentStatus( Status::PAY_RECEIVED );
+		}
+
+		return $response;
+	}
+
+	protected function getPaymentForm( \Aimeos\MShop\Order\Item\Iface $order, array $params ) : \Aimeos\MShop\Common\Helper\Form\Iface
```

Comment:
The whole method seems to be merged from the original getPaymentForm() and processOrder() methods but I'm sure this is wrong because the processOrder() code should be called when the payment has been finished and the status is confirmed by the gateway.

Reply:
answer of this issue is related to previous issue , I will return to it after solving previous one 
Regards